### PR TITLE
Email insights

### PR DESCRIPTION
**Pull request by richardolsson**
## Description
This PR implements an Insights page for emails, presenting key metrics for the user to assess the success of their email and compare it with other emails sent previously.

### Open rate graphs
image

### Click rate gauge and table
image

### Video
https://github.com/zetkin/app.zetkin.org/assets/550212/c7bb5175-e3ac-4ac8-a717-c0f330fb4b93

## Changes
* Adds new "Insights" tab to email page
* Creates an RPC to load insights data
* Renders information about email opens:
  * A gauge for the open rate, including optionally comparing to another email
  * A line chart of open rate over time, with support for optionally comparing to another email
  * A brief description of what "opens" mean and how to improve
* Renders information about clicks
  * A gauge for the click rate measured as CTR or CTOR, including optionally comparing to another email
  * A list of all the links in the email and how many times they have been clicked
  * A miniature of the email that highlights the links when hovering them in the list
  * A brief description of what CTR and CTOR mean, and how to improve them

## Notes to reviewer
This can't be tested on the dev server yet, because it requires a new API that hasn't been deployed yet. This feature does not make a huge sense for the dummy data that is on the dev server anyway, which is why I have tried it with real data during development.

## Related issues
Undocumented

### CODE REVIEW

**Comment by richardolsson** [comment]
Most of the day team is on vacation, so I would love to get some feedback on the code from @WULCAN or @henrycatalinismith if you get a chance. 🙏

I'm particularly curious about ideas to improve the big insights page component. What do you think would be a sensible way to split it up?

**Comment by henrycatalinismith** [comment]
How does it look if you extract these three into components? Loads of extra work to manage the increased surface area and shared state? Or just a bunch of prop loading in the page component and then they can each grab their precooked data and begin rendering real easy?

| | |
|-|-|
| Screenshot_2024-07-06_at_13_42_18 | Screenshot 2024-07-06 at 13 46 12 |

Personally become less and less dogmatic over the years about whether splitting stuff up into smaller components is strictly better or worse. But that's probably where I'd be looking first to split this one up once my mind was made up it was too big.

**Review by WULCAN — CHANGES_REQUESTED** [review]
I have reviewed all the changes, found only minor problems and added some questions and ideas for refactoring below.

I feel like `ZUIFutures` is a little problematic as it adds guarantees about the state of the future's data but those guarantees are not expressed to TypeScript. I imagine we might be able to resolve that by replacing `ZUIFutures` with `Suspense` but `use` is not yet in stable releases so it's probably too early.

https://react.dev/reference/react/Suspense
https://react.dev/reference/react/use

File: src/pages/organize/[orgId]/projects/[campId]/emails/[emailId]/insights.tsx
Comment (richardolsson) [reply src/pages/organize/[orgId]/projects/[campId]/emails/[emailId]/insights.tsx]: Good idea! 💯  Fixed in 5cfa7fa663def7bac81b2f47bdc0193cd9aff68f.

**Review by richardolsson — COMMENTED** [review]
Awesome work reviewing this. I've made lots of changes and replied to all(?) of your comments now. Wanna review again @WULCAN?

**Review by WULCAN — COMMENTED** [review]
I have some new comments, but none of them need to block merge. There are however a few unresolved threads that I still think are interesting before merge:

* https://github.com/zetkin/app.zetkin.org/pull/2076#discussion_r1679809340
* https://github.com/zetkin/app.zetkin.org/pull/2076#discussion_r1679975447
* https://github.com/zetkin/app.zetkin.org/pull/2076#discussion_r1685296688
* https://github.com/zetkin/app.zetkin.org/pull/2076#discussion_r1685367275
* https://github.com/zetkin/app.zetkin.org/pull/2076#discussion_r1685660186
* https://github.com/zetkin/app.zetkin.org/pull/2076#discussion_r1685669808
* https://github.com/zetkin/app.zetkin.org/pull/2076#discussion_r1685670578
* https://github.com/zetkin/app.zetkin.org/pull/2076#discussion_r1685672405
* https://github.com/zetkin/app.zetkin.org/pull/2076#discussion_r1685675887
* https://github.com/zetkin/app.zetkin.org/pull/2076#discussion_r1685680907
* https://github.com/zetkin/app.zetkin.org/pull/2076#discussion_r1685682388
* https://github.com/zetkin/app.zetkin.org/pull/2076#discussion_r1685672405
* https://github.com/zetkin/app.zetkin.org/pull/2076#discussion_r1729957004
* https://github.com/zetkin/app.zetkin.org/pull/2076#discussion_r1733375324

File: src/features/emails/components/ClickedInsightsSection.tsx
Comment (richardolsson) [reply src/features/emails/components/ClickedInsightsSection.tsx:1]: Well spotted! This is unfortunately a very common and easy-to-make mistake that I usually catch but this time didn't.

File: src/pages/organize/[orgId]/projects/[campId]/emails/[emailId]/insights.tsx
Comment (richardolsson) [reply src/pages/organize/[orgId]/projects/[campId]/emails/[emailId]/insights.tsx]: Because `getEmailInsights()` is code being run on the server (an RPC), and the code you're commenting on is being run on the client, so the data needs to be serializable between the two.

File: src/pages/organize/[orgId]/projects/[campId]/emails/[emailId]/insights.tsx
Comment (richardolsson) [reply src/pages/organize/[orgId]/projects/[campId]/emails/[emailId]/insights.tsx]: I don't fully understand your reasoning here, maybe because I'm not sure whether you've looked at the output of [nivo line charts](https://nivo.rocks/line/) with the various options (your reference to the underlying d3 library makes me think that maybe you haven't).

If I understand correctly, you are arguing that we should not use `basis` because the line will not perfectly hit the control points. I think that is a fair argument, if the goal is to perfectly illustrate the change from point to point. But trying to do so in these extremely intense diagrams will most likely produce very jagged lines, which (IMO) distracts from the more important attribute of the chart which is to visualize the trend.

File: src/zui/ZUICleanHtml.tsx
Comment (richardolsson) [reply src/zui/ZUICleanHtml.tsx:16]: You are correct that `Box` will re-render whenever `ZUICleanHtml` re-renders. Is that a problem in your opinion? It won't change the DOM though.

Personally I'm not entirely sure how `dangerouslySetInnerHTML` is implemented internally in React, but I don't expect the DOM to change if the HTML value did not change, so re-rendering the box should be entirely negligible.

The goal of using `useMemo()` here is not to avoid re-rendering, but to avoid re-sanitizing the HTML (which is a heavier operation).

File: src/pages/organize/[orgId]/projects/[campId]/emails/[emailId]/insights.tsx
Comment (richardolsson) [reply src/pages/organize/[orgId]/projects/[campId]/emails/[emailId]/insights.tsx]: Good point!

File: src/pages/organize/[orgId]/projects/[campId]/emails/[emailId]/insights.tsx
Comment (richardolsson) [reply src/pages/organize/[orgId]/projects/[campId]/emails/[emailId]/insights.tsx]: I don't think that's necessary. I'm a fan of duplicating once, and abstracting on the third occurrence (if the goal is DRY). But your comment made me realize another issue with this code which is that the entire return value from this function should be broken out into a separate component. Not for DRY, but for readability. So I'm going to do that! 😊

**Comment by richardolsson** [comment]
> I have some new comments, but none of them need to block merge. There are however a few unresolved threads that I still think are interesting before merge:

Thank you! I didn't understand why I didn't see some of these comments, but I'm realizing now that it's likely because you have been reviewing every commit individually and commenting on code that no longer exists in the PR, so when I go through all the comments on the current code I don't see those comments.

At least with my workflow, that is not a great way of receiving reviews. I understand the value of going through all the commits to understand the product (and the process of getting there), but in my opinion the review should be of the final product.

File: src/features/emails/components/EmailKPIChart.tsx
Comment (richardolsson) [reply src/features/emails/components/EmailKPIChart.tsx]: The `x` is the name of the radial bar, which I never intended to show so what is used here is purely for internal (code) use.

File: src/features/emails/components/EmailKPIChart.tsx
Comment (richardolsson) [reply src/features/emails/components/EmailKPIChart.tsx]: This comment was left on code that was no longer there in the final state of the PR when submitted for review, so I didn't see the comment at first. But to answer your question, yes, we are using two different series where one series is only used on the inner circle, and one is only used on the outer. The unused series have hard-coded zero as their value.

File: src/features/emails/components/EmailKPIChart.tsx
Comment (richardolsson) [reply src/features/emails/components/EmailKPIChart.tsx]: I try to avoid abstracting things until the third occurrence, or when necessary for other reasons (e.g. legibility).

File: src/features/emails/l10n/messageIds.ts
Comment (richardolsson) [reply src/features/emails/l10n/messageIds.ts]: This code had been removed by the time the PR was submitted for review.

File: src/pages/organize/[orgId]/projects/[campId]/emails/[emailId]/insights.tsx
Comment (richardolsson) [reply src/pages/organize/[orgId]/projects/[campId]/emails/[emailId]/insights.tsx:119]: You're right about the documentation. I don't know if this has changed in recent versions of React but I was almost 100% sure that `null` and `undefined` was treated differently.

Anyway, none of it applies here because we're not dealing (directly) with an `input` or `textarea` or whatever here. We're dealing with a MUI `Autocomplete` which follows different rules. I have tested to verify that `undefined` and `null` are treated differently and that `null` is fine (generates no warnings) while `undefined` is not (generates warning below).

image

File: src/features/emails/components/EmailKPIChart.tsx
Comment (richardolsson) [reply src/features/emails/components/EmailKPIChart.tsx]: I agree it would be nicer, and since it's quite a simple change to names that will still make sense, I'm making the change!

File: src/pages/organize/[orgId]/projects/[campId]/emails/[emailId]/insights.tsx
Comment (richardolsson) [reply src/pages/organize/[orgId]/projects/[campId]/emails/[emailId]/insights.tsx]: I prefer `null` to indicate an explicit non-value, whereas `undefined` is an implicit lack of value. Most languages don't make this distinction so I appreciate that this is completely arbitrary, but there are som practical conseqeuces.

For example, if we require a property to be `null` and not `undefined`, that requires the user of the component to pass the property with a `null` value, whereas if we allowed it to be `undefined` the property could be omitted altogether. This is probably why I have this feeling of one being "explicit" and the other "implicit".

However, in this case, I realized the `EmailKPIChart` properties did actually allow `undefined` which I think is wrong, so I've made the change now.

File: src/pages/organize/[orgId]/projects/[campId]/emails/[emailId]/insights.tsx
Comment (richardolsson) [reply src/pages/organize/[orgId]/projects/[campId]/emails/[emailId]/insights.tsx]: I've found that it works quite well, but even if it didn't reliably render 7 ticks (which I acknowledge that it won't always do) I don't see why that's a problem. The ticks will still render with labels that indicate the number of days so even if they are "weird" they will still be correct and the hover card will provide the necessary detail.

File: src/zui/ZUIDuration/index.tsx
Comment (richardolsson) [reply src/zui/ZUIDuration/index.tsx:51]: Yeah, concatenating sentences is bad practice. But I guess I don't think of this as quite the same thing. Here we are not really dealing with a sentence but with scalars and their units. It might still be a problem in some language that is very different than the languages we currently support, but I don't expect it to be.

File: src/pages/organize/[orgId]/projects/[campId]/emails/[emailId]/insights.tsx
Comment (richardolsson) [reply src/pages/organize/[orgId]/projects/[campId]/emails/[emailId]/insights.tsx]: Good point, and although I hadn't even considered it when I wrote this code, I do believe this is doing the right thing for what we want to show. We don't technically want to show the open rate at every point in time, but the rate at which the email reaches the eyes of the total target group.

File: src/pages/organize/[orgId]/projects/[campId]/emails/[emailId]/insights.tsx
Comment (richardolsson) [reply src/pages/organize/[orgId]/projects/[campId]/emails/[emailId]/insights.tsx]: Fixed in 2edc4c521cbd68fea28bd76d76afdb217cbdeb5a

**Comment by richardolsson** [comment]
I have made more changes now and am ready for another review. I appreciate the thoroughness @WULCAN!

**Review by WULCAN — COMMENTED** [review]
Thank you. It's a pleasure to read your changes. Here's a few more notes I wrote while reviewing the last changes. There's only one thread left where I want to learn more but I expect I will Approve the code as is.

File: src/pages/organize/[orgId]/projects/[campId]/emails/[emailId]/insights.tsx
Comment (richardolsson) [reply src/pages/organize/[orgId]/projects/[campId]/emails/[emailId]/insights.tsx]: You are completely right, both that we don't have local maximums/minimums, and that jaggedness is less of a problem for an ever-increasing line. But I still find that `monotoneX` is unnecessarily "jagged". I'm attaching two screenshots to show what I mean. So I think `basis` is actually fine (and better) for this.

## With `monotoneX`
image

## With `basis`
image

File: src/features/emails/components/EmailKPIChart.tsx
Comment (richardolsson) [reply src/features/emails/components/EmailKPIChart.tsx]: You have linked the same commit twice. I believe the second commit that you intended to reference is 2fdd0dd48d635e05cd21bda72aa9bb62dcf146e9.

I will try to explain this from memory without going back and experimenting and hopefully I remember my reasoning correctly.

The nivo radial bar allows for multiple "series", where each series will be rendered as a separate circle (radial bar). Each circle (radial bar) can have several segments, similarly to how a normal bar chart can have "stacked" bars. If the segments do not total 100% the remainder of the circle will be filled with the `tracksColor`.

Our use case is not 100% what radial bars were intended for, for two reasons:

1. Because we wish to render what is essentially the same data point ("open rate") in both circles, but in different colors. Normally, values for the same data point would have the same color across all series.
2. Because we wish to render a percentage, so a single value of 5% should not fill the entire circle, even if it's the only value (and thus constitutes 100% of the maximum)

These didn't cause big problems when I had just one series (before the comparison implemented in 2fdd0dd48d635e05cd21bda72aa9bb62dcf146e9) because 1 is only relevant for multiple series and 2 could be solved using `maxValue`, but when I have two emails the `maxValue` is actually different for each email and can no longer be used.

So what I've ended up doing is having one data point for the open rate of the primary email (which will be zero for the secondary email) and one data point for the open rate of the secondary email (which will be zero for the primary), and then finally one data point for the unopened.

File: src/features/emails/components/EmailDiagramHoverCard.tsx
Comment (richardolsson) [reply src/features/emails/components/EmailDiagramHoverCard.tsx:14]: You're right about all of these, and it might be possible to express stricter types as unions of the various states. I think it sounds great to explore it!

For the states that concern loading, I would actually suggest a bigger discussion about how we should be loading data. The `IFuture` path that we chose to explore (inspired by Dart/Flutter) has proven to have problems in React, and it might be better to completely shift to [`Suspense`](https://react.dev/reference/react/Suspense). This has been on the horizon for a while, which is why you're seeing a lot of "sloppiness" with loading states.

File: src/features/emails/components/EmailDiagramHoverCard.tsx
Comment (richardolsson) [reply src/features/emails/components/EmailDiagramHoverCard.tsx:85]: You're right! The main reason I changed it was because `numSent` is an attribute of the return value from `useEmailStats()` which is not the same as `mainStats` (which is actually `useEmailStats().data`).  But you're right that it also solves a divide-by-zero problem!

File: src/pages/organize/[orgId]/projects/[campId]/emails/[emailId]/insights.tsx
Comment (WULCAN) [reply src/pages/organize/[orgId]/projects/[campId]/emails/[emailId]/insights.tsx]: If the data is jagged, I think it's better to present at jagged line, unless you are trying to model some other underlying metric which is reasonably smooth.

Can you produce a third comparison screenshot with just the data points, without interpolation?

Comment (richardolsson) [reply src/pages/organize/[orgId]/projects/[campId]/emails/[emailId]/insights.tsx]: > If the data is jagged, I think it's better to present at jagged line, unless you are trying to model some other underlying metric which is reasonably smooth.

I am trying to indicate a trend, not visualize exact values at exact times. It could be argued then that the trend could be made visible by sampling the data at even intervals (e.g. one point per hour), but instead I have taken the easier route of presenting all data but with a smoothed line. I hope it's clear from the screenshots above that the choice between `monotoneX` and `basis` is, in the end, mostly one of personal taste.

> Can you produce a third comparison screenshot with just the data points, without interpolation?

Grabbing a new screenshot is actually a fair bit of work. I feel like this a matter of such detail that I'd rather not spend the time doing that.

**Review by WULCAN — APPROVED** [review]

**Pull request merged**
